core/model/modx/modfilehandler.class.php

```diff
@@ -514,6 +514,70 @@ public function create($mode = '') {
     }
 
     /**
+     * Iterates over a modDirectory object and returns an array of all containing files,
+     * optionally runs recursive, filters by file extension(s) and sorts the resulting list according to a specified sort flag
+     *
+     * @param array $options Options for iterating the directory.
+     *      boolean recursive If also subfolders should be scanned for files
+     *      boolean ignorehidden If folders and files starting with a dot . (hidden files/folders in unix envirionments) should be ignored, defaults to true
+     *      boolean sort If the resulting filelist array should be sorted with the specified flag like SORT_ASC, SORT_DESC, SORT_REGULAR, SORT_NATURAL, SORT_NUMERIC
+     *      string extensions Comma separated list of file extensions to filter files by
+     *      
+     * @return array of filepaths
+     */
+    public function getFiles($options = array()) {
+        $options = array_merge(array(
+            'recursive' => false,
+            'ignorehidden' => true,
+            'sort' => false,
+            'extensions' => ''
+        ), $options);
+
+        $files = array();
+        $extensions = explode(',', $options['extensions']);
+        
+        if ($options['recursive']) {
+            $iterator = new RecursiveDirectoryIterator($this->path);
+
+            foreach (new RecursiveIteratorIterator($iterator) as $file) {
+                $dirname = explode(DIRECTORY_SEPARATOR, $file->getPath()); // PHP Strict warning if we put end() around here
+                
+                if (($options['ignorehidden'] ? (substr($file->getFilename(), 0, 1) !== '.' && substr(end($dirname), 0, 1) !== '.') : true) && $file->isFile()) {
```

Comment:
=)...lol, definitely "choosing conciseness over readability" here...but how to avoid quadruplicating the following:

```php
           if (!empty($options['extensions'])) {
                            if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                                $files[] = $file->getPathname();
                            }
                        } else {
                            $files[] = $file->getPathname();
                        }
```
because if I take out the ugly line that checks everything then I have to wrap if / else statements around, which could look something like that, that is also not very pretty to me:

```php
                if ($options['ignorehidden'] && $file->isFile()) {

                    if (substr($file->getFilename(), 0, 1) !== '.' && substr(end($dirname), 0, 1) !== '.') {
                        if (!empty($options['extensions'])) {
                            if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                                $files[] = $file->getPathname();
                            }
                        } else {
                            $files[] = $file->getPathname();
                        }
                    }
                } else {
                    if (!empty($options['extensions'])) {
                        if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                            $files[] = $file->getPathname();
                        }
                    } else {
                        $files[] = $file->getPathname();
                    }                    
                }

```
did you mean something like that @Mark-H 

another option of keeping the tenary and avoid the code-quadruplication (it's already duplicate, that's ugly enough =D) would be to "extract" the long/ugly part into a variable, which would make the tenary a bit less ugly, e.g. the full code would look like that

```php
    public function getFiles($options = array()) {
        $options = array_merge(array(
            'recursive' => false,
            'ignorehidden' => true,
            'sort' => false,
            'extensions' => ''
        ), $options);
        $files = array();
        $extensions = explode(',', $options['extensions']);
        
        if ($options['recursive']) {
            $iterator = new RecursiveDirectoryIterator($this->path);
            foreach (new RecursiveIteratorIterator($iterator) as $file) {
                $dirname = explode(DIRECTORY_SEPARATOR, $file->getPath()); // PHP Strict warning if we put end() around here
                // would return true if first char in filename or dirname is a dot (.)
                $ishidden = substr($file->getFilename(), 0, 1) === '.' && substr(end($dirname), 0, 1) === '.';

                if (($options['ignorehidden'] ? !$ishidden : true) && $file->isFile()) {
                    if (!empty($options['extensions'])) {
                        if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                            $files[] = $file->getPathname();
                        }
                    } else {
                        $files[] = $file->getPathname();
                    }
                }
            }
        } else {
            $iterator = new DirectoryIterator($this->path);
            foreach ($iterator as $file) {
                // would return true if first char in filename is a dot (.)
                $ishidden = substr($file->getFilename(), 0, 1) === '.';

                if (($options['ignorehidden'] ? !$ishidden : true) && $file->isFile()) {
                    if (!empty($options['extensions'])) {
                        if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                            $files[] = $file->getPathname();
                        }
                    } else {
                        if ($file->isFile()) {
                            $files[] = $file->getPathname();
                        }
                    }
                }
            }
        }
        if (!empty($options['sort'])) {
            array_multisort($files, $options['sort'], $files);
        }
        return $files;
    }
```